File: fsharp/Problem11.fs
module Problem11

open System.Collections.Generic
open System.Diagnostics
open Utils

type Operation =
    | Add of uint64
    | Multiply of uint64
    | Square


let execute a  = function
    | Add b -> a + b
    | Multiply b -> a * b 
    | Square -> a * a


let swap<'a> (f: 'a -> 'a -> 'a) (a: 'a) (b: 'a) =
    f b a


let try_parse_suffix<'a> (prefix: string) (f: string -> 'a) (s: string) =
    if s.StartsWith prefix then Some (f (s.Substring prefix.Length))
    else None


let parse_suffix<'a> (prefix: string) (f: string -> 'a) (s: string) =
    match try_parse_suffix prefix f s with
    | Some result -> result
    | None -> failwith "Unrecognized prefix"
    

let parse_operation description =
    if description = "old * old" then Square
    else match try_parse_suffix "old * " uint64 description with
         | Some n -> Multiply n
         | None -> parse_suffix "old + " (uint64 >> Add) description
         
         
let rec gcd a b =
    if a < b then gcd b a
    else if b = 0UL then a
    else gcd b (a % b)


let inline lcm a b = a * b / (gcd a b)


type Monkey(items: uint64 seq,
            operation: Operation,
            divisible_by: uint64,
            if_true: int,
            if_false: int) =

    let items = Queue<uint64> items

    let mutable total_inspects = 0
    
    member _.Items = items :> uint64 seq
    
    member _.Operation = operation
    
    member _.DivisibleBy = divisible_by
    
    member _.IfTrue = if_true
    
    member _.IfFalse = if_false
    
    member _.TotalInspects = total_inspects
    
    member _.Queue = items |> Seq.map string |> String.concat ", "
    
    member _.AddItem item = items.Enqueue item
    
    member _.InspectAndThrow (monkeys: Monkey array) (relax: uint64 -> uint64) =
        total_inspects <- total_inspects + items.Count
        while items.Count > 0 do
            let item = items.Dequeue ()
            let worry_level = relax (execute item operation)
            let whom = if worry_level % divisible_by = 0UL
                       then if_true
                       else if_false
            
            monkeys[whom].AddItem worry_level
            
            
type Monkeys(monkeys: Monkey seq) =
    let monkeys = Array.ofSeq monkeys
    let common_lcm = monkeys |> Seq.map (fun it -> it.DivisibleBy)
                             |> Seq.scan lcm 1UL
                             |> Seq.last

    let mutable round = 0
    
    member _.Round = round
    
    member _.Items = monkeys :> Monkey seq
    
    member _.Inspects = monkeys |> Seq.map (fun it -> it.TotalInspects)
                                |> Seq.map string
                                |> String.concat ", "
                                
    member _.Queues = monkeys |> Seq.map (fun it -> it.Queue)
                              |> String.concat "; "
                                
    member _.Round1 () =
        for i in 0..monkeys.Length - 1 do
          monkeys[i].InspectAndThrow monkeys ((swap (/)) 3UL)

        round <- round + 1

    member _.Round2 () =
        for i in 0..monkeys.Length - 1 do
          monkeys[i].InspectAndThrow monkeys ((swap (%)) common_lcm)
          
        round <- round + 1
        

let parse_monkey (lines: string array) =
    Debug.Assert (lines.Length = 6)
    let items = parse_suffix "  Starting items: " (fun s -> s.Split(", ") |> Seq.map uint64) lines[1]
    let operation = parse_suffix "  Operation: new = " parse_operation lines[2]
    let divisible_by = parse_suffix "  Test: divisible by " uint64 lines[3]
    let if_true = parse_suffix "    If true: throw to monkey " int lines[4]
    let if_false = parse_suffix "    If false: throw to monkey " int lines[5]
    Monkey(items, operation, divisible_by, if_true, if_false)


let parse (lines: string seq) =
    lines |> Seq.split ""
          |> Seq.map Seq.toArray
          |> Seq.map parse_monkey
          |> Seq.toArray
          |> Monkeys


let solve_a (lines: string seq) =
    let monkeys = parse lines
    for _ in 1..20 do
        monkeys.Round1 ()
    
    monkeys.Items |> Seq.map (fun it -> it.TotalInspects |> uint64)
                  |> Seq.sortDescending
                  |> Seq.take 2
                  |> Seq.fold (*) 1UL
            
            
let solve_b (lines: string seq) =
    let monkeys = parse lines
    for _ in 1..10000 do
        monkeys.Round2 ()
        
    monkeys.Items |> Seq.map (fun it -> it.TotalInspects |> uint64)
                  |> Seq.sortDescending
                  |> Seq.take 2
                  |> Seq.fold (*) 1UL

File: rust/src/problem11.rs
use std::collections::vec_deque::VecDeque;
use std::fmt::Debug;
use std::str::FromStr;
use num::integer::lcm;

enum Operation {
    Add(u64),
    Multiply(u64),
    Square
}

impl Operation {
    fn apply(&self, b: u64) -> u64 {
        match &self {
            Operation::Add(a) => a + b,
            Operation::Multiply(a) => a * b,
            Operation::Square => b * b,
        }
    }
}

fn parse_operation(s: &str) -> Operation {
    if s == "  Operation: new = old * old" {
        Operation::Square
    } else if let Some(number) = s.strip_prefix("  Operation: new = old * ") {
        Operation::Multiply(number.parse::<u64>().unwrap())
    } else if let Some(number) = s.strip_prefix("  Operation: new = old + ") {
        Operation::Add(number.parse::<u64>().unwrap())
    } else {
        panic!("Unrecognized operation");
    }
}

fn parse_items(s: &str) -> VecDeque<u64> {
    if let Some(value) = s.strip_prefix("  Starting items: ") {
        value.split(", ").map(|s| s.parse::<u64>().unwrap()).collect()
    } else {
        panic!("Unrecognized items")
    }
}

fn parse_number<N>(s: &str, prefix: &str) -> N where N: FromStr, <N as FromStr>::Err: Debug {
    if let Some(value) = s.strip_prefix(prefix) {
        return N::from_str(value).unwrap();
    }
    
    panic!("Unrecognized number");
}

struct Monkey {
    items: VecDeque<u64>,
    operation: Operation,
    divisible_by: u64,
    if_true: usize,
    if_false: usize,
    total_inspects: usize,
}

impl Monkey {
    fn load(lines: &mut dyn Iterator<Item=String>) -> Self {
        debug_assert!(lines.next().unwrap().starts_with("Monkey"));
        let items = parse_items(&lines.next().unwrap());
        let operation = parse_operation(&lines.next().unwrap());
        let divisible_by = parse_number::<u64>(&lines.next().unwrap(), "  Test: divisible by ");
        let if_true = parse_number::<usize>(&lines.next().unwrap(), "    If true: throw to monkey ");
        let if_false = parse_number::<usize>(&lines.next().unwrap(), "    If false: throw to monkey ");
        let total_inspects = 0;

        Monkey { items, operation, divisible_by, if_true, if_false, total_inspects }
    }

    fn inspect<F>(&mut self, relax: F) -> Vec<(usize, u64)>  where F: Fn(u64) -> u64 {
        let mut result = Vec::new();
        self.total_inspects += self.items.len();
        while self.items.len() > 0 {
            let item = self.items.pop_front().unwrap();
            let worry_level = relax(self.operation.apply(item));
            let whom =
                if worry_level % self.divisible_by == 0 { self.if_true }
                else { self.if_false};

            result.push((whom, worry_level));
        }

        result
    }
}

struct Monkeys {
    monkeys: Vec<Monkey>,
    common_lcm: u64
}

impl Monkeys {
    fn load(lines: &mut dyn Iterator<Item=String>) -> Monkeys {
        let mut monkeys = vec![];

        loop {
            let monkey = Monkey::load(lines);
            monkeys.push(monkey);

            if lines.next() == None {
                break;
            }
        }

        let mut common_lcm = 1;
        for monkey in monkeys.iter() {
            common_lcm = lcm(common_lcm, monkey.divisible_by);
        }

        Monkeys { monkeys, common_lcm }
    }

    fn round1(&mut self) {
        let len = self.monkeys.len();
        for i in 0..len {
            let inspects = (&mut self.monkeys[i]).inspect(|a| a / 3u64);

            for (whom, worry_level) in inspects.iter() {
                (&mut self.monkeys[*whom]).items.push_back(*worry_level);
            }
        }
    }

    fn round2(&mut self) {
        let len = self.monkeys.len();
        let common_lcm = self.common_lcm;
        for i in 0..len {
            let inspects = (&mut self.monkeys[i]).inspect(|a| a % common_lcm);

            for (whom, worry_level) in inspects.iter() {
                (&mut self.monkeys[*whom]).items.push_back(*worry_level);
            }
        }
    }

    fn get_total_inspects(&self) -> Vec<usize> {
        self.monkeys.iter().map(|monkey| monkey.total_inspects).collect()
    }
}

pub fn solve_a(lines: &mut dyn Iterator<Item=String>) -> usize {
    let mut monkeys = Monkeys::load(lines);
    for _ in 0..20 {
        monkeys.round1();
    }

    let mut inspects = monkeys.get_total_inspects();
    inspects.sort_by(|a, b| b.cmp(a));
    inspects.iter()
        .take(2)
        .product()
}

#[test]
fn solve_a_with_sample_data_returns_10605() {
    let sample = indoc::indoc!("
        Monkey 0:
          Starting items: 79, 98
          Operation: new = old * 19
          Test: divisible by 23
            If true: throw to monkey 2
            If false: throw to monkey 3
        
        Monkey 1:
          Starting items: 54, 65, 75, 74
          Operation: new = old + 6
          Test: divisible by 19
            If true: throw to monkey 2
            If false: throw to monkey 0
        
        Monkey 2:
          Starting items: 79, 60, 97
          Operation: new = old * old
          Test: divisible by 13
            If true: throw to monkey 1
            If false: throw to monkey 3
        
        Monkey 3:
          Starting items: 74
          Operation: new = old + 3
          Test: divisible by 17
            If true: throw to monkey 0
            If false: throw to monkey 1");
    let mut lines = crate::utils::str_to_iter(sample);

    let actual = solve_a(&mut lines);

    assert_eq!(10_605, actual);
}

pub fn solve_b(lines: &mut dyn Iterator<Item=String>) -> usize {
    let mut monkeys = Monkeys::load(lines);
    for _ in 0..10000 {
        monkeys.round2();
    }

    let mut inspects = monkeys.get_total_inspects();
    inspects.sort_by(|a, b| b.cmp(a));
    inspects.iter()
        .take(2)
        .product()
}

#[ignore]
#[test]
fn solve_b_with_sample_data_returns_2713310158() {
    let sample = indoc::indoc!("
        Monkey 0:
          Starting items: 79, 98
          Operation: new = old * 19
          Test: divisible by 23
            If true: throw to monkey 2
            If false: throw to monkey 3
        
        Monkey 1:
          Starting items: 54, 65, 75, 74
          Operation: new = old + 6
          Test: divisible by 19
            If true: throw to monkey 2
            If false: throw to monkey 0
        
        Monkey 2:
          Starting items: 79, 60, 97
          Operation: new = old * old
          Test: divisible by 13
            If true: throw to monkey 1
            If false: throw to monkey 3
        
        Monkey 3:
          Starting items: 74
          Operation: new = old + 3
          Test: divisible by 17
            If true: throw to monkey 0
            If false: throw to monkey 1");
    let mut lines = crate::utils::str_to_iter(sample);

    let actual = solve_a(&mut lines);

    assert_eq!(2_713_310_158_usize, actual);
}
